AX: For every clearChildren-addChildren cycle, we create new AccessibilityImageMapLink instances and never clean them up, causing a memory leak

twilco opened the pull request:
#### eed31338f551e7934e2a8901262a1238cb8b1142
<pre>
AX: For every clearChildren-addChildren cycle, we create new AccessibilityImageMapLink instances and never clean them up, causing a memory leak
<a href="https://bugs.webkit.org/show_bug.cgi?id=284125">https://bugs.webkit.org/show_bug.cgi?id=284125</a>
<a href="https://rdar.apple.com/141013232">rdar://141013232</a>

Reviewed by Joshua Hoffman.

Like most other things generated by AXObjectCache::create(AccessibilityRole), AccessibilityImageMapLinks are never
are not persisted between clear-children add-children cycles, and are never passed in a call to AXObjectCache::remove,
meaning they leak forever.

Fix this by making AccessibilityImageMapLink subclass AccessibilityNodeObject, inherently tying their lifetime to the
HTMLAreaElement that they represent.

This commit also eliminates redundant role ImageMapLink. There is no reason for this role to exist over other existing
roles, WebCoreLink and Link.

* LayoutTests/accessibility/image-map-title-causes-crash-expected.txt:
* LayoutTests/accessibility/image-map-title-causes-crash.html:
* LayoutTests/imported/w3c/web-platform-tests/html-aam/fragile/area-role-expected.txt:
* LayoutTests/platform/glib/accessibility/image-map2-expected.txt:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::isLink const):
(WebCore::AXCoreObject::isImplicitlyInteractive const):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::accessibilityRoleToString):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::focusedImageMapUIElement):
(WebCore::AXObjectCache::createFromNode):
(WebCore::AXObjectCache::getOrCreate):
(WebCore::AXObjectCache::create):
* Source/WebCore/accessibility/AccessibilityImageMapLink.cpp:
(WebCore::AccessibilityImageMapLink::AccessibilityImageMapLink):
(WebCore::AccessibilityImageMapLink::create):
(WebCore::AccessibilityImageMapLink::parentObject const):
(WebCore::AccessibilityImageMapLink::determineAccessibilityRole):
(WebCore::AccessibilityImageMapLink::anchorElement const):
(WebCore::AccessibilityImageMapLink::url const):
(WebCore::AccessibilityImageMapLink::imageMapLinkRenderer const):
(WebCore::AccessibilityImageMapLink::elementPath const):
(WebCore::AccessibilityImageMapLink::elementRect const):
(WebCore::AccessibilityImageMapLink::setHTMLAreaElement): Deleted.
(WebCore::AccessibilityImageMapLink::detachFromParent): Deleted.
* Source/WebCore/accessibility/AccessibilityImageMapLink.h:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::documentLinks):
(WebCore::AccessibilityRenderObject::addImageMapChildren):
(WebCore::AccessibilityRenderObject::addChildren):
* Source/WebCore/accessibility/atspi/AccessibilityAtspi.cpp:
* Source/WebCore/accessibility/atspi/AccessibilityObjectAtspi.cpp:
(WebCore::atspiRole):
* Source/WebCore/accessibility/cocoa/AXCoreObjectCocoa.mm:
(WebCore::Accessibility::createPlatformRoleMap):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper accessibilityCanFuzzyHitTest]):
(-[WebAccessibilityObjectWrapper determineIsAccessibilityElement]):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(children):

Canonical link: <a href="https://commits.webkit.org/293278@main">https://commits.webkit.org/293278@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4d8c7ea231b698b91624de8cdba9aa98d5be26e7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97992 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17623 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7850 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103107 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48521 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100037 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17915 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26074 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74607 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-view-transitions/active-view-transition-type-on-non-root.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31798 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100996 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13573 "Found 1 new test failure: fast/forms/ios/focus-input-via-button.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88554 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54967 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13353 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6484 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47963 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83335 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6567 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/105485 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25078 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18264 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83595 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/25451 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84732 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83047 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21059 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27691 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5367 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18701 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25037 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/30211 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24857 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28173 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/26432 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fleizach commented on Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
```diff
@@ -2618,8 +2606,12 @@ void AccessibilityRenderObject::addChildren()
-                addChildIfNeeded(*childObject);
+            // <area> elements should only be added via addImageMapChildren(), as their expected hierarchy
+            // differs from the structure of the DOM.
+            if (LIKELY(!is<HTMLAreaElement>(child))) {
```
is there a way to avoid this special case at this point in adding children? is this the only element type that needs to do this special behavior?

twilco replied:
I don't think there's a way to avoid a special-case somewhere, since we have explicitly chosen to provide a better user experience by exposing a different accessibility tree structure than what is in the DOM.

But you are right that we probably need this for other object types too — you could imagine that we would want this for `display:contents` elements too. So I moved the logic into `AccessibilityObject::insertChild`, which we funnel everything through. Thanks for the catch!

fleizach commented on Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
```diff
@@ -1079,6 +1079,8 @@ - (NSBezierPath *)path
 
 static NSArray *children(AXCoreObject& backingObject)
 {
+    // FIXME: This function calls unignored children multiple times, which is inefficient.
```
Can we just replace the usage on line 1098 with the original call to unignored children?

twilco replied:
Yes! I've done so in the latest commit.

hoffmanjoshua commented on Source/WebCore/accessibility/AccessibilityObject.cpp:
```diff
@@ -635,6 +635,18 @@ void AccessibilityObject::insertChild(AccessibilityObject& child, unsigned index
     }
 
+    if (UNLIKELY(is<HTMLAreaElement>(child.node()))) {
+        // Despite the DOM parent for <area> elements being <map>, we expose <area> elements as children
```
Appreciate this explanation here!

twilco committed:
AX: For every clearChildren-addChildren cycle, we create new AccessibilityImageMapLink instances and never clean them up, causing a memory leak
https://bugs.webkit.org/show_bug.cgi?id=284125
rdar://141013232

Reviewed by Joshua Hoffman.

Like most other things generated by AXObjectCache::create(AccessibilityRole), AccessibilityImageMapLinks are never
are not persisted between clear-children add-children cycles, and are never passed in a call to AXObjectCache::remove,
meaning they leak forever.

Fix this by making AccessibilityImageMapLink subclass AccessibilityNodeObject, inherently tying their lifetime to the
HTMLAreaElement that they represent.

This commit also eliminates redundant role ImageMapLink. There is no reason for this role to exist over other existing
roles, WebCoreLink and Link.

* LayoutTests/accessibility/image-map-title-causes-crash-expected.txt:
* LayoutTests/accessibility/image-map-title-causes-crash.html:
* LayoutTests/imported/w3c/web-platform-tests/html-aam/fragile/area-role-expected.txt:
* LayoutTests/platform/glib/accessibility/image-map2-expected.txt:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::isLink const):
(WebCore::AXCoreObject::isImplicitlyInteractive const):
* Source/WebCore/accessibility/AXCoreObject.h:
(WebCore::accessibilityRoleToString):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::focusedImageMapUIElement):
(WebCore::AXObjectCache::createFromNode):
(WebCore::AXObjectCache::getOrCreate):
(WebCore::AXObjectCache::create):
* Source/WebCore/accessibility/AccessibilityImageMapLink.cpp:
(WebCore::AccessibilityImageMapLink::AccessibilityImageMapLink):
(WebCore::AccessibilityImageMapLink::create):
(WebCore::AccessibilityImageMapLink::parentObject const):
(WebCore::AccessibilityImageMapLink::determineAccessibilityRole):
(WebCore::AccessibilityImageMapLink::anchorElement const):
(WebCore::AccessibilityImageMapLink::url const):
(WebCore::AccessibilityImageMapLink::imageMapLinkRenderer const):
(WebCore::AccessibilityImageMapLink::elementPath const):
(WebCore::AccessibilityImageMapLink::elementRect const):
(WebCore::AccessibilityImageMapLink::setHTMLAreaElement): Deleted.
(WebCore::AccessibilityImageMapLink::detachFromParent): Deleted.
* Source/WebCore/accessibility/AccessibilityImageMapLink.h:
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::documentLinks):
(WebCore::AccessibilityRenderObject::addImageMapChildren):
(WebCore::AccessibilityRenderObject::addChildren):
* Source/WebCore/accessibility/atspi/AccessibilityAtspi.cpp:
* Source/WebCore/accessibility/atspi/AccessibilityObjectAtspi.cpp:
(WebCore::atspiRole):
* Source/WebCore/accessibility/cocoa/AXCoreObjectCocoa.mm:
(WebCore::Accessibility::createPlatformRoleMap):
* Source/WebCore/accessibility/ios/WebAccessibilityObjectWrapperIOS.mm:
(-[WebAccessibilityObjectWrapper accessibilityCanFuzzyHitTest]):
(-[WebAccessibilityObjectWrapper determineIsAccessibilityElement]):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(children):

Canonical link: https://commits.webkit.org/293278@main
